gapis/resolve/delete.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (C) 2019 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resolve
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/gapid/core/memory/arena"
+	"github.com/google/gapid/gapis/api"
+	"github.com/google/gapid/gapis/capture"
+	"github.com/google/gapid/gapis/database"
+	"github.com/google/gapid/gapis/service/path"
+)
+
+// Delete creates a copy of the capture referenced by p, but without the object, value
+// or memory at p. The path returned is identical to p, but with
+// the base changed to refer to the new capture.
+func Delete(ctx context.Context, p *path.Any, r *path.ResolveConfig) (*path.Any, error) {
+	obj, err := database.Build(ctx, &DeleteResolvable{Path: p, Config: r})
+	if err != nil {
+		return nil, err
+	}
+
+	return obj.(*path.Any), nil
+}
+
+// Resolve implements the database.Resolver interface.
+func (r *DeleteResolvable) Resolve(ctx context.Context) (interface{}, error) {
+	ctx = SetupContext(ctx, path.FindCapture(r.Path.Node()), r.Config)
+
+	p, err := deleteCommand(ctx, arena.New(), r.Path.Node())
+	if err != nil {
+		return nil, err
+	}
+
+	return p.Path(), nil
+}
+
+func deleteCommand(ctx context.Context, a arena.Arena, p path.Node) (*path.Capture, error) {
+	switch p := p.(type) {
+	case *path.Command:
+		if len(p.Indices) > 1 {
+			return nil, fmt.Errorf("Cannot modify subcommands") // TODO: Subcommands
+		}
+
+		cmdIdx := p.Indices[0]
+
+		// Resolve the command list
+		oldCmds, err := NCmds(ctx, p.Capture, cmdIdx+1)
+		if err != nil {
+			return nil, err
+		}
+
+		cmds := trimCommandList(cmdIdx, oldCmds)
+
+		// Create the new capture
+		old, err := capture.ResolveGraphicsFromPath(ctx, p.Capture)
+		if err != nil {
+			return nil, err
+		}
+
+		c, err := capture.NewGraphicsCapture(ctx, a, old.Name()+"*", old.Header, old.InitialState, cmds)
+		if err != nil {
+			return nil, err
+		}
+
+		newCapure, err := capture.New(ctx, c)
+		if err != nil {
+			return nil, err
+		}
+
+		return newCapure, nil
+	}
+	return nil, fmt.Errorf("Incorrect path type %T", p)
+}
+
+func trimCommandList(cmdIdx uint64, oldCmds []api.Cmd) []api.Cmd {
+	const MAXID = ^uint(0)
+
+	cmds := make([]api.Cmd, 0, len(oldCmds)-1)
+	cmds = append(cmds, oldCmds[:cmdIdx]...)
+	cmds = append(cmds, oldCmds[cmdIdx+1:]...)
+
+	var childrenToRemove []uint64
+
+	for i, cmd := range cmds {
+		callerID := cmd.Caller()
+
+		if callerID == api.CmdNoID {
+			continue
+		}
+
+		if uint64(callerID) == cmdIdx {
```

Comment:
We may have a similar issue with command ID stored in Resources, e.g.:
https://github.com/google/gapid/blob/f8e98b1c4d728586569d2f92770b390b0bdd5d83/gapis/service/service.proto#L967

@AWoloszyn what do you think? This command index updating may cascade in many places?

Reply:
Another point here: can we assume that for a command of index i and its callerID of index j, we always have i < j ? (i.e., the callerID is always of greater index than the command itself?)

If so, then I think we can avoid the recursion here and just update command indexes in one sweep over `oldCmds[cmdIdx+1:]`, because changing the ID may only affect commands that are later in the trace (but that's off the top of my head, I may have missed some details).